src/main/java/carpet/script/annotation/ListConverter.java

```diff
@@ -0,0 +1,90 @@
+package carpet.script.annotation;
+
+import java.lang.reflect.AnnotatedParameterizedType;
+import java.lang.reflect.AnnotatedType;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import carpet.script.value.ListValue;
+import carpet.script.value.Value;
+
+/**
+ * <p>Converts a given {@link ListValue} into a {@link List} of values converted to {@code <T>}.</p>
+ * 
+ * <p>If the {@link Param.AllowSingleton} annotation is specified, allows creating a singleton from
+ * a loose element compatible with the type conversion.</p> 
+ * 
+ * <p>Lists provided by this converter are <b>not</b> linked to the initial list, and therefore will not
+ * reflect changes in either of them</p>
+ *
+ * @param <T> The type of the element that will be inside the list
+ */
+final class ListConverter<T> implements ValueConverter<List<T>>
+{
+    private final ValueConverter<T> itemConverter;
+    private final boolean allowSingletonCreation;
+
+    @Override
+    public String getTypeName()
+    {
+        return (allowSingletonCreation ? itemConverter.getTypeName() + " or " : "") + "list of " + itemConverter.getTypeName() + "s";
+    }
+
+    @Override
+    public List<T> convert(Value value)
+    {
+        return value instanceof ListValue ? convertListValue((ListValue) value) : allowSingletonCreation ? convertSingleton(value) : null;
+    }
+
+    private List<T> convertListValue(ListValue values)
+    {
+        List<T> list = new ArrayList<>(values.getItems().size());
+        for (Value value : values)
+        {
+            T converted = itemConverter.convert(value);
+            if (converted == null)
+                return null;
+            list.add(converted);
+        }
+        return list;
+    }
+
+    private List<T> convertSingleton(Value val)
+    {
+        T converted = itemConverter.convert(val);
+        if (converted == null)
+            return null;
+        return Collections.singletonList(converted);
```

Comment:
Should singletons be returned as normal ArrayLists as well? Aren't they non-mutable? I have caught myself returning .asList() for things like block positions and someone complained you can't alter the size of these. For that reason I replaced all listvalue construtors to use new ArrayList.